<h1 align='center'>
 <b> EDA </b>
</h1>


# Yellow Taxis
<p align='center'> 
Trip Records:

In [1]:
import pandas as pd

Importation and reading of both yellows taxis trips

In [4]:
df_yel22= pd.read_parquet('yellow_tripdata_2022-06.parquet')
df_yel23= pd.read_parquet('yellow_tripdata_2023-06.parquet')

<p align='center'> <b> 2022 :</b>

In [5]:
df_yel22.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3558124 entries, 0 to 3558123
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int64         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int64         
 8   DOLocationID           int64         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  airport_fee           

Here we see that there are no string data in any column, we just have numberic, date and time values in this dataset.

The following is a sample from thw dataframe of the 3558124 trips drived by yellow taxis in NYC during the month of June in 2022

In [6]:
df_yel22.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2022-06-01 00:25:41   2022-06-01 00:48:22              1.0   
1         1  2022-06-01 00:44:40   2022-06-01 01:01:48              1.0   
2         2  2022-06-01 00:23:07   2022-06-01 00:39:50              1.0   
3         1  2022-06-01 00:25:53   2022-06-01 00:57:06              2.0   
4         1  2022-06-01 00:23:58   2022-06-01 00:33:43              0.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0          11.00         1.0                  N            70            48   
1           4.20         1.0                  N           170           226   
2           9.49         1.0                  N           264           113   
3          12.10         1.0                  N           132            17   
4           1.80         1.0                  N           140           163   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1         32.0   3.00      0.5        2.00          6.55   
1             1         14.0   3.00      0.5        0.00          0.00   
2             1         26.0   0.50      0.5        5.00          6.55   
3             2         37.0   1.75      0.5        0.00          0.00   
4             1          9.0   3.00      0.5        2.55          0.00   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    0.3         44.35                   2.5         0.00  
1                    0.3         17.80                   2.5         0.00  
2                    0.3         42.60                   2.5         1.25  
3                    0.3         39.55                   0.0         1.25  
4                    0.3         15.35                   2.5         0.00

Missing and Null data:
-

Below is shown that there are five columns with 132448 null values in each.

Incomplete Columns:
- passenger_count
- RatecodeID
- store_and_fwd_flag
- congestion_surcharge
- airport_fee

In [7]:
df_yel22.isnull().sum()

VendorID                      0
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count          132448
trip_distance                 0
RatecodeID               132448
store_and_fwd_flag       132448
PULocationID                  0
DOLocationID                  0
payment_type                  0
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge     132448
airport_fee              132448
dtype: int64

Duplicated rows:
-

There are 0 duplicated rows in this dataset

In [9]:
df_yel22.duplicated().sum()

0

In [11]:
df_yel22.columns

Index(['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
       'passenger_count', 'trip_distance', 'RatecodeID', 'store_and_fwd_flag',
       'PULocationID', 'DOLocationID', 'payment_type', 'fare_amount', 'extra',
       'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge',
       'total_amount', 'congestion_surcharge', 'airport_fee'],
      dtype='object')

List of the Columns without ID's columns (but VendedorID):

In [32]:
column=['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime','passenger_count', 'trip_distance', 'store_and_fwd_flag', 'fare_amount', 'extra','mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge','total_amount', 'congestion_surcharge', 'airport_fee']

In [33]:
df_yel22[column]

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2022-06-01 00:25:41   2022-06-01 00:48:22              1.0   
1               1  2022-06-01 00:44:40   2022-06-01 01:01:48              1.0   
2               2  2022-06-01 00:23:07   2022-06-01 00:39:50              1.0   
3               1  2022-06-01 00:25:53   2022-06-01 00:57:06              2.0   
4               1  2022-06-01 00:23:58   2022-06-01 00:33:43              0.0   
...           ...                  ...                   ...              ...   
3558119         1  2022-06-30 23:45:51   2022-06-30 23:51:48              NaN   
3558120         2  2022-06-30 23:25:00   2022-06-30 23:40:00              NaN   
3558121         2  2022-06-30 23:29:00   2022-06-30 23:37:00              NaN   
3558122         2  2022-06-30 23:24:15   2022-06-30 23:50:19              NaN   
3558123         2  2022-06-30 23:33:53   2022-06-30 23:54:58              NaN   

         trip_distance store_and_fwd_flag  fare_amount  extra  mta_tax  \
0                11.00                  N        32.00   3.00      0.5   
1                 4.20                  N        14.00   3.00      0.5   
2                 9.49                  N        26.00   0.50      0.5   
3                12.10                  N        37.00   1.75      0.5   
4                 1.80                  N         9.00   3.00      0.5   
...                ...                ...          ...    ...      ...   
3558119           0.00               None         9.20   0.50      0.5   
3558120           5.01               None        18.86   0.00      0.5   
3558121           1.55               None        10.03   0.00      0.5   
3558122           5.30               None        24.34   0.00      0.5   
3558123           4.41               None        21.16   0.00      0.5   

         tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0              2.00          6.55                    0.3         44.35   
1              0.00          0.00                    0.3         17.80   
2              5.00          6.55                    0.3         42.60   
3              0.00          0.00                    0.3         39.55   
4              2.55          0.00                    0.3         15.35   
...             ...           ...                    ...           ...   
3558119        0.00          0.00                    0.3         15.00   
3558120        5.19          0.00                    0.3         27.35   
3558121        3.10          0.00                    0.3         16.43   
3558122        0.00          0.00                    0.3         27.64   
3558123        0.00          0.00                    0.3         24.46   

         congestion_surcharge  airport_fee  
0                         2.5         0.00  
1                         2.5         0.00  
2                         2.5         1.25  
3                         0.0         1.25  
4                         2.5         0.00  
...                       ...          ...  
3558119                   NaN          NaN  
3558120                   NaN          NaN  
3558121                   NaN          NaN  
3558122                   NaN          NaN  
3558123                   NaN          NaN  

[3558124 rows x 15 columns]

The following function shows the Count number of non-NA/null observations, the Mean of the values, the Maximum of the values in column, the Minimum of the values in column, the standard deviation of each column data, and the 25% ,50% and 75%.

In [15]:
df_yel22[columns].describe()

VendorID        tpep_pickup_datetime       tpep_dropoff_datetime  \
count  3.558124e+06                     3558124                     3558124   
mean   1.714216e+00  2022-06-14 23:44:14.353117  2022-06-15 06:45:32.857034   
min    1.000000e+00         2001-08-23 05:34:45         2001-08-23 05:57:11   
25%    1.000000e+00         2022-06-08 14:01:07  2022-06-08 14:21:32.750000   
50%    2.000000e+00  2022-06-15 18:13:03.500000         2022-06-15 18:32:14   
75%    2.000000e+00  2022-06-23 10:25:20.250000         2022-06-23 10:43:03   
max    6.000000e+00         2023-04-18 14:30:05         2023-04-18 23:30:39   
std    4.877039e-01                         NaN                         NaN   

       passenger_count  trip_distance   fare_amount         extra  \
count     3.425676e+06   3.558124e+06  3.558124e+06  3.558124e+06   
mean      1.399042e+00   5.968216e+00  1.524926e+01  1.021005e+00   
min       0.000000e+00   0.000000e+00 -9.070000e+02 -7.000000e+00   
25%       1.000000e+00   1.140000e+00  7.000000e+00  0.000000e+00   
50%       1.000000e+00   1.930000e+00  1.050000e+01  5.000000e-01   
75%       1.000000e+00   3.670000e+00  1.700000e+01  2.500000e+00   
max       9.000000e+00   3.070071e+05  3.958449e+05  8.250000e+00   
std       9.607126e-01   5.941291e+02  2.121836e+02  1.254616e+00   

            mta_tax    tip_amount  tolls_amount  improvement_surcharge  \
count  3.558124e+06  3.558124e+06  3.558124e+06           3.558124e+06   
mean   4.884199e-01  2.795843e+00  5.622156e-01           2.960561e-01   
min   -5.000000e-01 -8.008000e+01 -6.320000e+01          -3.000000e-01   
25%    5.000000e-01  1.000000e+00  0.000000e+00           3.000000e-01   
50%    5.000000e-01  2.160000e+00  0.000000e+00           3.000000e-01   
75%    5.000000e-01  3.450000e+00  0.000000e+00           3.000000e-01   
max    3.300000e+00  1.400160e+03  8.000900e+02           3.000000e-01   
std    9.378860e-02  3.581685e+00  2.121173e+00           4.810925e-02   

       total_amount  congestion_surcharge   airport_fee  
count  3.558124e+06          3.425676e+06  3.425676e+06  
mean   2.211842e+01          2.282375e+00  9.573366e-02  
min   -9.115500e+02         -2.500000e+00 -1.250000e+00  
25%    1.235000e+01          2.500000e+00  0.000000e+00  
50%    1.618000e+01          2.500000e+00  0.000000e+00  
75%    2.334000e+01          2.500000e+00  0.000000e+00  
max    3.958482e+05          2.750000e+00  1.250000e+00  
std    2.124623e+02          7.490437e-01  3.357336e-01

<p align='center'> <b> 2023 :</b>

In [17]:
df_yel23.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3307234 entries, 0 to 3307233
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int32         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int32         
 8   DOLocationID           int32         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  Airport_fee           

Here we see that there are no string data in any column, we just have numberic, date and time values in this dataset.

The following is a sample from the dataframe of the 3307234 trips drived by yellow taxis in NYC during the month of June in 2023

In [16]:
df_yel23.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2023-06-01 00:08:48   2023-06-01 00:29:41              1.0   
1         1  2023-06-01 00:15:04   2023-06-01 00:25:18              0.0   
2         1  2023-06-01 00:48:24   2023-06-01 01:07:07              1.0   
3         2  2023-06-01 00:54:03   2023-06-01 01:17:29              3.0   
4         2  2023-06-01 00:18:44   2023-06-01 00:27:18              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           3.40         1.0                  N           140           238   
1           3.40         1.0                  N            50           151   
2          10.20         1.0                  N           138            97   
3           9.83         1.0                  N           100           244   
4           1.17         1.0                  N           137           234   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1         21.9   3.50      0.5        6.70           0.0   
1             1         15.6   3.50      0.5        3.00           0.0   
2             1         40.8   7.75      0.5       10.00           0.0   
3             1         39.4   1.00      0.5        8.88           0.0   
4             1          9.3   1.00      0.5        0.72           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  Airport_fee  
0                    1.0         33.60                   2.5         0.00  
1                    1.0         23.60                   2.5         0.00  
2                    1.0         60.05                   0.0         1.75  
3                    1.0         53.28                   2.5         0.00  
4                    1.0         15.02                   2.5         0.00

Missing and Null data:
-

Below is shown that there are five columns with 99887 null values in each.

Incomplete Columns:
- passenger_count
- RatecodeID
- store_and_fwd_flag
- congestion_surcharge
- airport_fee

In [18]:
df_yel23.isnull().sum()

VendorID                     0
tpep_pickup_datetime         0
tpep_dropoff_datetime        0
passenger_count          99887
trip_distance                0
RatecodeID               99887
store_and_fwd_flag       99887
PULocationID                 0
DOLocationID                 0
payment_type                 0
fare_amount                  0
extra                        0
mta_tax                      0
tip_amount                   0
tolls_amount                 0
improvement_surcharge        0
total_amount                 0
congestion_surcharge     99887
Airport_fee              99887
dtype: int64

Duplicated rows:
-

There are 0 duplicated rows in this dataset

In [21]:
df_yel23.duplicated().sum()

0

In [26]:
df_yel23.columns

Index(['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
       'passenger_count', 'trip_distance', 'RatecodeID', 'store_and_fwd_flag',
       'PULocationID', 'DOLocationID', 'payment_type', 'fare_amount', 'extra',
       'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge',
       'total_amount', 'congestion_surcharge', 'Airport_fee'],
      dtype='object')

List of the Columns without ID's columns (but VendedorID):

In [35]:
column= ['VendorID', 'tpep_pickup_datetime', 'tpep_dropoff_datetime','passenger_count', 'trip_distance', 'store_and_fwd_flag', 'payment_type', 'fare_amount', 'extra','mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge','total_amount', 'congestion_surcharge', 'Airport_fee']

In [36]:
df_yel23[column]

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2023-06-01 00:08:48   2023-06-01 00:29:41              1.0   
1               1  2023-06-01 00:15:04   2023-06-01 00:25:18              0.0   
2               1  2023-06-01 00:48:24   2023-06-01 01:07:07              1.0   
3               2  2023-06-01 00:54:03   2023-06-01 01:17:29              3.0   
4               2  2023-06-01 00:18:44   2023-06-01 00:27:18              1.0   
...           ...                  ...                   ...              ...   
3307229         2  2023-06-30 23:30:21   2023-07-01 00:00:46              NaN   
3307230         1  2023-06-30 23:34:22   2023-07-01 00:32:59              NaN   
3307231         2  2023-06-30 23:45:00   2023-07-01 00:10:00              NaN   
3307232         2  2023-06-30 23:13:38   2023-06-30 23:51:19              NaN   
3307233         2  2023-06-30 23:18:00   2023-06-30 23:31:00              NaN   

         trip_distance store_and_fwd_flag  payment_type  fare_amount  extra  \
0                 3.40                  N             1        21.90   3.50   
1                 3.40                  N             1        15.60   3.50   
2                10.20                  N             1        40.80   7.75   
3                 9.83                  N             1        39.40   1.00   
4                 1.17                  N             1         9.30   1.00   
...                ...                ...           ...          ...    ...   
3307229           9.20               None             0        38.53   0.00   
3307230          20.20               None             0        70.00   1.75   
3307231           6.16               None             0        28.03   0.00   
3307232          11.65               None             0        45.74   0.00   
3307233           1.40               None             0        14.35   0.00   

         mta_tax  tip_amount  tolls_amount  improvement_surcharge  \
0            0.5        6.70          0.00                    1.0   
1            0.5        3.00          0.00                    1.0   
2            0.5       10.00          0.00                    1.0   
3            0.5        8.88          0.00                    1.0   
4            0.5        0.72          0.00                    1.0   
...          ...         ...           ...                    ...   
3307229      0.5        8.51          0.00                    1.0   
3307230      0.5       11.97          6.55                    1.0   
3307231      0.5        6.41          0.00                    1.0   
3307232      0.5        0.00          6.55                    1.0   
3307233      0.5        0.00          0.00                    1.0   

         total_amount  congestion_surcharge  Airport_fee  
0               33.60                   2.5         0.00  
1               23.60                   2.5         0.00  
2               60.05                   0.0         1.75  
3               53.28                   2.5         0.00  
4               15.02                   2.5         0.00  
...               ...                   ...          ...  
3307229         51.04                   NaN          NaN  
3307230         91.77                   NaN          NaN  
3307231         38.44                   NaN          NaN  
3307232         56.29                   NaN          NaN  
3307233         18.35                   NaN          NaN  

[3307234 rows x 16 columns]

The following function shows the Count number of non-NA/null observations, the Mean of the values, the Maximum of the values in column, the Minimum of the values in column, the standard deviation of each column data, and the 25% ,50% and 75%.

In [37]:
df_yel23[column].describe()

VendorID        tpep_pickup_datetime       tpep_dropoff_datetime  \
count  3.307234e+06                     3307234                     3307234   
mean   1.737262e+00  2023-06-15 18:09:27.394469  2023-06-15 18:27:09.958314   
min    1.000000e+00         2002-12-31 23:03:19         2002-12-31 23:06:17   
25%    1.000000e+00  2023-06-08 09:12:42.250000  2023-06-08 09:30:06.250000   
50%    2.000000e+00  2023-06-15 14:15:24.500000         2023-06-15 14:36:09   
75%    2.000000e+00         2023-06-23 02:25:20  2023-06-23 02:40:05.750000   
max    6.000000e+00         2023-07-01 00:42:13         2023-07-03 16:31:24   
std    4.477026e-01                         NaN                         NaN   

       passenger_count  trip_distance  payment_type   fare_amount  \
count     3.207347e+06   3.307234e+06  3.307234e+06  3.307234e+06   
mean      1.369012e+00   4.368790e+00  1.187126e+00  1.998804e+01   
min       0.000000e+00   0.000000e+00  0.000000e+00 -7.500000e+02   
25%       1.000000e+00   1.080000e+00  1.000000e+00  9.300000e+00   
50%       1.000000e+00   1.820000e+00  1.000000e+00  1.350000e+01   
75%       1.000000e+00   3.540000e+00  1.000000e+00  2.260000e+01   
max       9.000000e+00   1.852505e+05  5.000000e+00  3.869836e+05   
std       8.959539e-01   2.280398e+02  5.478269e-01  2.136687e+02   

              extra       mta_tax    tip_amount  tolls_amount  \
count  3.307234e+06  3.307234e+06  3.307234e+06  3.307234e+06   
mean   1.620755e+00  4.859043e-01  3.594915e+00  6.065670e-01   
min   -7.500000e+00 -5.000000e-01 -9.500000e+01 -7.255000e+01   
25%    0.000000e+00  5.000000e-01  1.000000e+00  0.000000e+00   
50%    1.000000e+00  5.000000e-01  2.860000e+00  0.000000e+00   
75%    2.500000e+00  5.000000e-01  4.480000e+00  0.000000e+00   
max    1.425000e+01  5.750000e+00  7.118800e+02  6.655600e+02   
std    1.869549e+00  1.068704e-01  4.203569e+00  2.252982e+00   

       improvement_surcharge  total_amount  congestion_surcharge   Airport_fee  
count           3.307234e+06  3.307234e+06          3.207347e+06  3.207347e+06  
mean            9.805643e-01  2.906873e+01          2.262730e+00  1.534284e-01  
min            -1.000000e+00 -7.510000e+02         -2.500000e+00 -1.750000e+00  
25%             1.000000e+00  1.596000e+01          2.500000e+00  0.000000e+00  
50%             1.000000e+00  2.130000e+01          2.500000e+00  0.000000e+00  
75%             1.000000e+00  3.130000e+01          2.500000e+00  0.000000e+00  
max             1.000000e+00  3.869876e+05          2.500000e+00  1.750000e+00  
std             1.941525e-01  2.141569e+02          7.960739e-01  5.027511e-01

</h1>

# For-Hire Vehicule
<p align='center'> 
Trip Records:

Importation and reading of both yellows taxis trips

In [38]:
df_fhv22= pd.read_parquet('fhv_tripdata_2022-06.parquet')
df_fhv23= pd.read_parquet('fhv_tripdata_2023-06.parquet')

<p align='center'> <b> 2022 :</b>

In [39]:
df_fhv22.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1195414 entries, 0 to 1195413
Data columns (total 7 columns):
 #   Column                  Non-Null Count    Dtype         
---  ------                  --------------    -----         
 0   dispatching_base_num    1195414 non-null  object        
 1   pickup_datetime         1195414 non-null  datetime64[us]
 2   dropOff_datetime        1195414 non-null  datetime64[us]
 3   PUlocationID            335321 non-null   float64       
 4   DOlocationID            1043505 non-null  float64       
 5   SR_Flag                 0 non-null        object        
 6   Affiliated_base_number  1195414 non-null  object        
dtypes: datetime64[us](2), float64(2), object(3)
memory usage: 63.8+ MB


The following is a sample from thw dataframe of the 1195414 trips drived by for-hire vehicle in NYC during the month of June in 2022

In [40]:
df_fhv22.head()

dispatching_base_num     pickup_datetime    dropOff_datetime  PUlocationID  \
0               B00013 2022-06-01 00:22:00 2022-06-01 01:18:00           NaN   
1               B00013 2022-06-01 00:54:24 2022-06-01 01:20:00           NaN   
2               B00014 2022-06-01 00:30:00 2022-06-01 01:12:00           NaN   
3               B00037 2022-06-01 00:18:51 2022-06-01 00:21:24           NaN   
4               B00037 2022-06-01 00:34:17 2022-06-01 00:40:12           NaN   

   DOlocationID SR_Flag Affiliated_base_number  
0           NaN    None                 B00013  
1           NaN    None                 B00013  
2           NaN    None                 B00013  
3          72.0    None                 B00037  
4          35.0    None                 B00037

Missing and Null data:
-

Below is shown that there are three columns with null values data.

The <b>SR_Flag</b> column is completly incomplete, all celds are empty.

Incomplete Columns:
- PUlocationID (with 860093 null values)
- DOlocationID (with 151909 null values)

In [41]:
df_fhv22.isnull().sum()

dispatching_base_num            0
pickup_datetime                 0
dropOff_datetime                0
PUlocationID               860093
DOlocationID               151909
SR_Flag                   1195414
Affiliated_base_number          0
dtype: int64

Duplicated rows:
-

In [42]:
df_fhv22.duplicated().sum()

1030

In [45]:
1030*100/1195414

0.08616261813898783

There are 1030 duplicated rows in this dataset, the 0.086%.

</h1>

In [47]:
df_fhv22.columns

Index(['dispatching_base_num', 'pickup_datetime', 'dropOff_datetime',
       'PUlocationID', 'DOlocationID', 'SR_Flag', 'Affiliated_base_number'],
      dtype='object')

The following function shows the Count number of non-NA/null observations, the Mean of the values, the Maximum of the values in column, the Minimum of the values in column, the standard deviation of each column data, and the 25% ,50% and 75%.

In [48]:
df_fhv22.describe()

pickup_datetime            dropOff_datetime   PUlocationID  \
count                     1195414                     1195414  335321.000000   
mean   2022-06-15 23:38:36.835069  2022-06-16 00:35:59.990386     135.944006   
min           2022-06-01 00:00:00         2022-06-01 00:01:19       1.000000   
25%    2022-06-08 12:43:23.500000  2022-06-08 13:12:11.250000      73.000000   
50%           2022-06-15 18:02:57         2022-06-15 18:46:21     137.000000   
75%    2022-06-23 11:22:23.750000         2022-06-23 11:53:01     206.000000   
max           2022-06-30 23:59:59         2022-12-06 23:56:00     265.000000   
std                           NaN                         NaN      75.333314   

       DOlocationID  
count  1.043505e+06  
mean   1.367917e+02  
min    1.000000e+00  
25%    6.900000e+01  
50%    1.330000e+02  
75%    2.130000e+02  
max    2.650000e+02  
std    7.966959e+01

<p align='center'> <b> 2023 :</b>

In [49]:
df_fhv23.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1219445 entries, 0 to 1219444
Data columns (total 7 columns):
 #   Column                  Non-Null Count    Dtype         
---  ------                  --------------    -----         
 0   dispatching_base_num    1219445 non-null  object        
 1   pickup_datetime         1219445 non-null  datetime64[us]
 2   dropOff_datetime        1219445 non-null  datetime64[us]
 3   PUlocationID            295894 non-null   float64       
 4   DOlocationID            1044549 non-null  float64       
 5   SR_Flag                 0 non-null        float64       
 6   Affiliated_base_number  1219445 non-null  object        
dtypes: datetime64[us](2), float64(3), object(2)
memory usage: 65.1+ MB


The following is a sample from thw dataframe of the 1195414 trips drived by for-hire vehicle in NYC during the month of June in 2023

In [50]:
df_fhv23.head()

dispatching_base_num     pickup_datetime    dropOff_datetime  PUlocationID  \
0               B00008 2023-06-01 00:00:00 2023-06-01 00:21:00           NaN   
1               B00008 2023-06-01 00:15:00 2023-06-01 00:41:00           NaN   
2               B00009 2023-06-01 00:00:00 2023-06-01 00:13:00           NaN   
3               B00009 2023-06-01 00:15:00 2023-06-01 00:28:00           NaN   
4               B00009 2023-06-01 00:30:00 2023-06-01 00:45:00           NaN   

   DOlocationID  SR_Flag Affiliated_base_number  
0           NaN      NaN                 B00008  
1           NaN      NaN                 B00008  
2           NaN      NaN                 B00009  
3           NaN      NaN                 B00009  
4           NaN      NaN                 B00009

Missing and Null data:
-

Below is shown that there are three columns with null values data.

The <b>SR_Flag</b> column is completly incomplete, all celds are empty.

Incomplete Columns:
- PUlocationID (with 923551 null values)
- DOlocationID (with 174896 null values)

In [51]:
df_fhv23.isnull().sum()

dispatching_base_num            0
pickup_datetime                 0
dropOff_datetime                0
PUlocationID               923551
DOlocationID               174896
SR_Flag                   1219445
Affiliated_base_number          0
dtype: int64

Duplicated rows:
-

In [53]:
df_fhv23.duplicated().sum()

624

In [54]:
624*100/1219445

0.05117081951215512

There are 1030 duplicated rows in this dataset, the 0.051%.

</h1>

In [56]:
df_fhv23.columns

Index(['dispatching_base_num', 'pickup_datetime', 'dropOff_datetime',
       'PUlocationID', 'DOlocationID', 'SR_Flag', 'Affiliated_base_number'],
      dtype='object')

The following function shows the Count number of non-NA/null observations, the Mean of the values, the Maximum of the values in column, the Minimum of the values in column, the standard deviation of each column data, and the 25% ,50% and 75%.

In [57]:
df_fhv23.describe()

pickup_datetime            dropOff_datetime   PUlocationID  \
count                     1219445                     1219445  295894.000000   
mean   2023-06-16 01:41:43.048444  2023-06-16 02:17:29.424864     139.453091   
min           2023-06-01 00:00:00         2023-06-01 00:03:17       1.000000   
25%           2023-06-08 10:22:55         2023-06-08 10:47:00      76.000000   
50%           2023-06-15 17:00:42         2023-06-15 17:43:00     141.000000   
75%           2023-06-23 11:54:10         2023-06-23 12:25:00     213.000000   
max           2023-06-30 23:59:59         2023-07-28 01:15:56     265.000000   
std                           NaN                         NaN      76.362197   

       DOlocationID  SR_Flag  
count  1.044549e+06      0.0  
mean   1.355564e+02      NaN  
min    1.000000e+00      NaN  
25%    6.900000e+01      NaN  
50%    1.330000e+02      NaN  
75%    2.120000e+02      NaN  
max    2.650000e+02      NaN  
std    7.925735e+01      NaN